### PR TITLE
Feature/first version

### PR DESCRIPTION
**Pull request by AnatoleMombrun**
# Goal of the PR
Have a first functional version of the nt user script which uses Excel for input and output, and which does the extraction fo the nt users info in parallel.

# Details
- First version of the tool working with an Excel document for input and ouput
- Excel path given in the YAML configuration file.
- Minimal Python version set to 3.8
- Parallel execution of 'net user' commands using ThreadPoolExecutor

# Not done in this PR:
- unit tests

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -2,12 +2,13 @@
 name = "nt-users-tool"
 version = "0.1.0"
 description = ""
-authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
+authors = ["Anatole Mombrun <anatole.mombrun@gmail.com>"]
```
Comment (LoicRiegel) [inline pyproject.toml]: Please revert this change. This field should indicate the owner /main maintainer of the repo, which is me.
You can look [here](https://github.com/eclipse/kiso-testing/blob/master/pyproject.toml) (the repo of the ITF project) to see that only the owner is listed here.
Your name will be seen on GitHub since you will be the author of some commits!

File: src/nt_users_tool/constants.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+from lib2to3.pgen2.token import NAME
```
Comment (LoicRiegel) [inline src/nt_users_tool/constants.py]: What is this?
Maybe this was generated by VS code and you didn't notice it?

File: src/nt_users_tool/excelprocessing.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+from openpyxl import Workbook, load_workbook
+from datetime import date
+
+from constants import *
```
Comment (LoicRiegel) [inline src/nt_users_tool/excelprocessing.py]: "Wildcard imports" (= `import *`) should be banned :)
Please use `from xxx import THIS, THAT` instead (so only what you need, and this is explicit)

Pleaes make this improvement for all the files :)

File: src/nt_users_tool/netcommands.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    """Generates a dictionnary with relevant user_information from nt_user_id via a net user shell command.
+
+    Args:
+        string_input (str): the nt_user_id for which the command will be executed
+
+    Returns:
+        dict: The dictionnary containing relevant information about this user
+    """
+    input = nt_user_id.upper()
+    command = 'net user /domain ' + input
```
Comment (LoicRiegel) [inline src/nt_users_tool/netcommands.py]: use f-string here?

File: src/nt_users_tool/netcommands.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import os
+
+from constants import *
+
+
+def obtain_info_from_user(nt_user_id: str):
```
Comment (LoicRiegel) [inline src/nt_users_tool/netcommands.py]: This function does too many things in my opinion.

What do you think about having those function?
```Python
def get_nt_user_info(nt_user:string) -> str
    """Returns the nt user info for the given NT user as a string by executing the 'net command /domain' command."""

def get_all_nt_users_info(nt_users:list[str]) -> list[str]:
    """Returns all nt users info in a list."""

def extract_nt_user_info(nt_user_info:str) -> NTUserInfo:
    """Extract NTUser info from a string and returns it as a NTUserInfo object.""" # NTUserInfo can be a typing.NamedTuple for instance

def extract_all_nt_user_info(nt_users_info:list[str]) -> list[NTUserInfo]:
    """Extract NTUser info from a lsit of nt users string and returns it as a list of NTUserInfo object."""
```

File: src/nt_users_tool/netcommands.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    """Generates a dictionnary with relevant user_information from nt_user_id via a net user shell command.
+
+    Args:
+        string_input (str): the nt_user_id for which the command will be executed
+
+    Returns:
+        dict: The dictionnary containing relevant information about this user
```
Comment (LoicRiegel) [inline src/nt_users_tool/netcommands.py]: A dictionary would work to store the relevant information about a user, but it's nice to be even more precise when defining types to clearly indicate what we expect. What are those relevant information about the user?

In my opinion, you can use a `NamedTuple` (in the typing module), simple and enough to do the job :)

File: src/nt_users_tool/yamlprocessing.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import yaml
+
+
+def output_file_path(config_file_name: str):
```
Comment (LoicRiegel) [inline src/nt_users_tool/yamlprocessing.py]: Currently, you open the file and read from it twice, which is not optimal. Here is what I suggest:
* have 1 function to read from the file and just return the content
* have 1 function to get a particular config from the content. Inside this function, we should handle the possibility that the config that we are currently trying to get is missing! In this case, let's throw an exception and exit the script right away.

What do you think?

File: src/nt_users_tool/excelprocessing.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    """_summary_
+
+    :param worksheet: _description_
+    :type worksheet: _type_
+    :param row_number: _description_
+    :param columns: _description_
+    :param user_info: _description_
+    """
+
+    keys = list(user_info.keys())
```
Comment (LoicRiegel) [inline src/nt_users_tool/excelprocessing.py]: `dict.keys()` already returns a list

File: src/nt_users_tool/excelprocessing.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+def nt_user_list(excel_file_name: str):
+    """Returns the list of nt_user IDs , reading them from the excel file.
+
+    Args:
+        excel_file_name (str): Excel Input filepath
+
+    Returns:
+        list: List of nt_user IDs
+    """
+    workbook = load_workbook(excel_file_name, read_only=True)
```
Comment (LoicRiegel) [inline src/nt_users_tool/excelprocessing.py]: I suggest to pass a workbook object to this function and do the `load_workbook` and `wb.close()` etc... in another function or in the main function.

File: src/nt_users_tool/excelprocessing.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    worksheet = workbook.active
```
Comment (LoicRiegel) [inline src/nt_users_tool/excelprocessing.py]: Here, you are assuming that when you open the Excel file for the first time, the active worksheet will always be the worksheet where the input of nt users are. I don't think that it assumption will be true all the time :) I would switch to a given worksheet, which name is given as config in the yaml file

File: src/nt_users_tool/constants.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+
+BLANK = " "
```
Comment (LoicRiegel) [inline src/nt_users_tool/constants.py]: Are you sure you need those constants?

File: src/nt_users_tool/constants.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+SLASH = "/"
+
+USERNAME = "user_name"
```
Comment (LoicRiegel) [inline src/nt_users_tool/constants.py]: okay, I think I understand those constants, let's see if we can do something better :) Let's discuss it

File: src/nt_users_tool/constants.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+LAST_RELEVANT_ELEMENT_POSITION = 40
+
+NT_USER_COLUMN = "D"
+NAME_COLUMN = "E"
+EXPIRATION_DATE_COLUMN = "F"
+
+COLUMNS_LIST = [NT_USER_COLUMN, NAME_COLUMN, EXPIRATION_DATE_COLUMN]
+
+FAKE_NET_COMMAND = "The request will be processed at a domain controller for domain de.bosch.com\
```
Comment (LoicRiegel) [inline src/nt_users_tool/constants.py]: those fake strings, you might be able to move them in the unit tests part when you're doing them. They might be useful for mocking

**Review by LoicRiegel — CHANGES_REQUESTED** [review]
Looks good, let's discuss in a meeting how we can improve things :)

File: src/nt_users_tool/yamlprocessing.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
+import yaml
+
```
Comment (LoicRiegel) [inline src/nt_users_tool/yamlprocessing.py]: Something more generic for the name of the module? "read_config.py" for instance?

BTW modules should follow the snake_case format. Can you apply that evverywhere please?

File: src/nt_users_tool/constants.py (hunk 2 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/constants.py]: Indeed. Conflicting with a custom variable named "NAME". Removed.

File: src/nt_users_tool/constants.py (hunk 11 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/constants.py]: Was added for "cleanliness" of the code but it is unnecessary. Will be removed.

File: src/nt_users_tool/excelprocessing.py (hunk 3 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excelprocessing.py]: Will be changed.

File: src/nt_users_tool/yamlprocessing.py (hunk 14 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/yamlprocessing.py]: File names will be changed.

File: src/nt_users_tool/yamlprocessing.py (hunk 7 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/yamlprocessing.py]: The current functions will be removed. Your solution is adopted.

File: src/nt_users_tool/netcommands.py (hunk 4 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/netcommands.py]: Syntax changed to f-string.

File: src/nt_users_tool/netcommands.py (hunk 6 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/netcommands.py]: Data format changed from dict to NamedTuple subclass NTUserInfo.

File: src/nt_users_tool/excelprocessing.py (hunk 8 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excelprocessing.py]: It does not exactly return a list (class dict_keys). I needed to access indexes hence the typecast.

File: src/nt_users_tool/netcommands.py (hunk 5 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/netcommands.py]: Changes made accordingly to suggestion.

File: src/nt_users_tool/excelprocessing.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    :param user_info: _description_
+    """
+
+    keys = list(user_info.keys())
+    for i, column in enumerate(columns):
+        print(column + str(row_number), keys[i], user_info[keys[i]])
+        worksheet[column + str(row_number)] = user_info[keys[i]]
+
+
+def nt_user_list(excel_file_name: str):
```
Comment (LoicRiegel) [inline src/nt_users_tool/excelprocessing.py]: `read_nt_users' ?

File: src/nt_users_tool/constants.py (hunk 13 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/constants.py]: Still not moved with commit 1ea2511, will move them later on.

File: src/nt_users_tool/excelprocessing.py (hunk 15 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excelprocessing.py]: Function name changed to read_nt_users for clarity.

File: src/nt_users_tool/excelprocessing.py (hunk 10 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excelprocessing.py]: We no longer use excel path as an imput for this function. The worksheet name will be located in the config file.

File: src/nt_users_tool/excelprocessing.py (hunk 9 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excelprocessing.py]: It will be done in the main() function.

File: config.yml [hunk 16]
```diff
@@ -0,0 +1,2 @@
+input_file_path: ../input.xlsx
```
Comment (LoicRiegel) [inline config.yml]: Don't forget that the name of the input nt user sheet must also be given here :)

Also, I think that in practice, the `input_file_path` and `output_file_path` will be the same. But we can keep 2 separate config settings.

File: pyproject.toml [hunk 17]
```diff
@@ -1,13 +1,14 @@
-[tool.poetry]
++[tool.poetry]
```
Comment (LoicRiegel) [inline pyproject.toml]: ⚠️ 
```suggestion
[tool.poetry]
```

File: pyproject.toml [hunk 18]
```diff
@@ -1,13 +1,14 @@
...
 name = "nt-users-tool"
 version = "0.1.0"
 description = ""
-authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
+authors = ["Loïc Rigel <loic.riegel@outlook.fr>"]
```
Comment (LoicRiegel) [inline pyproject.toml]: ⚠️ 
```suggestion
authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
```

File: src/nt_users_tool/excel_processing.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+def create_results_workbook(excel_desired_path: str, sheets_names_list: list):
+    """Creates the output excel file according to arguments
+
+    Args:
+        excel_desired_path (str): Path where the output file should be located
+        sheets_names_list (list): Names of the sheets inside the file
+    """
+    workbook = Workbook()
+    workbook.remove(workbook.active)
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: You might not want to do that! In case the output file is the same as the input file, and you just want to add /overwrite the ouptut worksheets

File: src/nt_users_tool/net_commands.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
+import os
+
+from typing import NamedTuple
+from constants import LAST_RELEVANT_ELEMENT_POSITION
+
+class NTUserInfo(NamedTuple):
```
Comment (LoicRiegel) [inline src/nt_users_tool/net_commands.py]: You can also add a docstring for the class. For is this data structure for? What data does it hold?

File: src/nt_users_tool/read_config.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import yaml
+from constants import CONFIG_FILE_PATH, INPUT_FILE, OUTPUT_FILE
+
+def read_config_file() -> dict():
```
Comment (LoicRiegel) [inline src/nt_users_tool/read_config.py]: ⚠️ do not create a new dictionary here!
```suggestion
def read_config_file() -> dict:
```

If you want to be more precise, you can also specify the type of the dict's keys and values. For instance:
```Python
from typing import Dict, Any
def read_config_file() -> Dict[str, Any]:
```
(`dict[str, Any]` works with Python >= 3.10)

**Review by LoicRiegel — CHANGES_REQUESTED** [review]

File: src/nt_users_tool/excel_processing.py [hunk 22]
```diff
@@ -0,0 +1,82 @@
+from openpyxl import Workbook, load_workbook
+from datetime import date
+
+from constants import *
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: Looks like you forgot this  `import *`  :)

File: pyproject.toml (hunk 17 above)
Comment (AnatoleMombrun) [reply pyproject.toml]: I also noticed it. It will be changed as the project no longer runs with this typo.

File: pyproject.toml (hunk 18 above)
Comment (AnatoleMombrun) [reply pyproject.toml]: Oops...

File: src/nt_users_tool/net_commands.py (hunk 20 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/net_commands.py]: Okay. Will add.

File: src/nt_users_tool/read_config.py (hunk 21 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/read_config.py]: Typo. I tried to apply the syntax you described but does not work with python 3.8.5 as you mentionned.

File: src/nt_users_tool/excel_processing.py (hunk 22 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: fixed in 04b3eac

File: src/nt_users_tool/excel_processing.py (hunk 19 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: Okay. This was done in case the output file was not the same as the input.

**Comment by LoicRiegel** [comment]
I switched the destination branch to `develop`

File: src/nt_users_tool/excelprocessing.py (hunk 3 above)
Comment (LoicRiegel) [reply src/nt_users_tool/excelprocessing.py]: I see you removed them, thanks

File: src/nt_users_tool/net_commands.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    expiration_date: str
+
+def get_nt_user_string(nt_user: str) -> str:
+    """Executes a net command with given nt_user and returns the response as a string.
+
+    :param nt_user: The nt_user that will be used in the net command.
+    :return: The response as a string.
+    """
+    input = nt_user.upper()
+    command = (f"net user /domain {input}")
```
Comment (LoicRiegel) [inline src/nt_users_tool/net_commands.py]: there is no need for brackets
```suggestion
    command = f"net user /domain {input}"
```

File: src/nt_users_tool/net_commands.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
+from os import popen
+from concurrent.futures import Executor, ThreadPoolExecutor
+
+from typing import NamedTuple
+
+from constants import LAST_RELEVANT_ELEMENT_POSITION
```
Comment (LoicRiegel) [inline src/nt_users_tool/net_commands.py]: Please use relative imports like this `from .constants ...` since we have a package. Can you please make the change in all imports?

We could also import from the package `from nt_user_tool.constants import ...`

File: src/nt_users_tool/yamlprocessing.py (hunk 14 above)
Comment (LoicRiegel) [reply src/nt_users_tool/yamlprocessing.py]: I see it's done, thanks

File: config.yml [hunk 25]
```diff
@@ -0,0 +1,2 @@
+input_file_path: ./input.xlsx
```
Comment (LoicRiegel) [inline config.yml]: let's name this file with ".yaml" instead

File: src/nt_users_tool/excel_processing.py [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+    :type worksheet: an openpyxl worksheet object.
+    :param row_number: the row at which the changes happen.
+    :param columns: the range of columns we write on.
+    :param user_info: The information that we are writing.
+    """
+    for i, column in enumerate(columns):
+        worksheet[column + str(row_number)] = nt_user_info[i]
+
+
+def fill_all_sheets(workbook: Workbook, nt_user_info_list: list):
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: here you can be more precise about the type hint
```Python
nt_user_info_list: list[NTUserInfo]
```
It helps the IDE understand what you are passing to the function which can then help you debug way more easilly :)

File: src/nt_users_tool/excel_processing.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+
+    all_user_index = 1
+    expired_index = 1
+    expiring_index = 1
+
+    for user in nt_user_info_list:
+        fill_one_row(users_sheet, all_user_index, COLUMNS_LIST, user)
+        all_user_index += 1
+
+        user_day, user_month, user_year = user.expiration_date.split("/")
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: Do you need to create those variables?
Could you use `user.expiration_date.day`, `user.expiration_date.month` etc... instaed?

File: src/nt_users_tool/excel_processing.py [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+        user_month = int(user_month)
+        user_year = int(user_year)
+        if user_year < now_year:
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: This logic is there to determine if a user is expired or not, or if he/she is expiring soon. This logic thus don't belong here in the excel processing file.
I suggest creating a new module called "nt_user.py" or "nt_user_info.py". Inside it, you will put the NTUserInfo class and some functions to compute if the user is expired or not.
I would recommand to use an enum for the expiration status (okay, expires soon, or expired).

In the top of the module, you can craete an object `date_now = date.now()`. This object will only be generated once, when python imports this module. You can then use this object inside the module. I am suggesting this so that we don't have to pass the current date as an argument to this function to determinte the expiration status. Let me know if what I 'm saying is clear aha :)

File: src/nt_users_tool/excel_processing.py [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+                expired_index += 1
+            elif user_month == now_month:
+                fill_one_row(expired_users_sheet, expired_index, COLUMNS_LIST, user)
+                expired_index +=1
+            elif user_month - now_month < MONTHS_TO_EXPIRE:
+                fill_one_row(expiring_soon_users_sheet, expiring_index, COLUMNS_LIST, user)
+                expiring_index +=1
+        elif user_year - now_year < TWO_YEAR_GAP:
+            if now_month > user_month:
+                if (user_month - now_month)%12 <= MONTHS_TO_EXPIRE:
```
Comment (LoicRiegel) [inline src/nt_users_tool/excel_processing.py]: Maybe you can simplify this logic. datetime.date objects can be compared easily
```Python
from datetime import date

d_after = date(2023, 2, 1)
d_before = date(2022, 12, 1)

elapsed = d_after - d_before
print("Nb days:", elapsed.days) # Will print 62
```

File: src/nt_users_tool/net_commands.py [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+
+class NTUserInfo(NamedTuple):
+    """Dataclass for a user of the bosch network
+
+    :param NamedTuple: Full name, nt_user and expiration date (fetched from net user /domain)
+    """
+    full_name: str
+    nt_user: str
+    expiration_date: str
```
Comment (LoicRiegel) [inline src/nt_users_tool/net_commands.py]: I suggest to use a more appropriate type for that, maybe datetime.date

File: src/nt_users_tool/net_commands.py [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    list_of_nt_user_string = []
+    for nt_user in list_nt_user:
+        list_of_nt_user_string.append(get_nt_user_string(nt_user))
+    return list_of_nt_user_string
+
+def extract_nt_user_info(nt_user_string: str) -> NTUserInfo:
+    """Extracts relevant data from a string generated with a given nt_user.
+
+    :param nt_user_string: Net command response as a string with given nt_user.
```
Comment (LoicRiegel) [inline src/nt_users_tool/net_commands.py]: Based on the description of this param, I would recommand naming it something like "nt_command_response"

**Review by LoicRiegel — COMMENTED** [review]
Thank you very much for fixing lots of things already! I put more comments, have fun 😂

File: src/nt_users_tool/net_commands.py (hunk 23 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/net_commands.py:24]: Brackets removed.

File: src/nt_users_tool/net_commands.py (hunk 24 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/net_commands.py:6]: Imports are now done with the package syntax nt_user_tool.module.

File: src/nt_users_tool/excel_processing.py (hunk 26 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: image

File: src/nt_users_tool/excel_processing.py (hunk 27 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py:62]: Those variables are created because when fetching from the net command, we obtain a date that looks like DD/MM/YYYY. I don't know how to write the syntax with expiration_date.month or .day. Could you please support me on that?

File: src/nt_users_tool/net_commands.py (hunk 30 above)
Comment (LoicRiegel) [reply src/nt_users_tool/net_commands.py]: great!

File: src/nt_users_tool/constants.py (hunk 13 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/constants.py]: Moved in ef87cb4.

File: config.yml (hunk 25 above)
Comment (AnatoleMombrun) [reply config.yml]: Fixed in 04d8101.

File: src/nt_users_tool/excel_processing.py (hunk 27 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: Fixed in 04d8101.

File: src/nt_users_tool/excel_processing.py (hunk 28 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: Fixed in 04d8101.

File: src/nt_users_tool/excel_processing.py (hunk 29 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: Fixed in 04d8101.

File: src/nt_users_tool/net_commands.py (hunk 31 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/net_commands.py]: Fixed in 04d8101.

File: src/nt_users_tool/excel_processing.py (hunk 26 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/excel_processing.py]: Fixed in 638ef90.

File: src/nt_users_tool/read_config.py (hunk 21 above)
Comment (AnatoleMombrun) [reply src/nt_users_tool/read_config.py]: Now fixed in 638ef90.

File: src/nt_users_tool/excel_processing.py (hunk 26 above)
Comment (LoicRiegel) [reply src/nt_users_tool/excel_processing.py]: yes, you can only do that in Python 3.10
With older versions you have to do
```Python
from typing import List
List[MyElement]
```

**Comment by LoicRiegel** [comment]
@AnatoleMombrun I just pushed new commits. You can have a look. I also updated the PR description.
Those are mainly minor fixes.

I also added a simple security check before doing the popen command to avoid code injection attacks.

Also, this project is public, so keep in mind that no sensible information should be pushed here.


Let's test the new commits tomorrow, and if the script works, we merge this PR :)

**Review by LoicRiegel — APPROVED** [review]
LGTM now, thank you a lot for this work!
